File: 2019/src/bin/day4.c
#include <stdio.h>
#include <stdbool.h>
#include <inttypes.h>

int main(int argc, char **argv)
{
	if (argc < 2) {
		fprintf(stderr, "Incorrect number of arguments provided\n");
		return 1;
	}
	FILE *f = fopen(argv[1], "r");
	if (f == NULL) {
		fprintf(stderr, "Could not open input file\n");
		return 0;
	}

	uint32_t start, end;
	fscanf(f, "%u-%u", &start, &end);
	fclose(f);

	uint32_t p1_count = 0;
	uint32_t p2_count = 0;
	for (uint32_t n = start; n <= end; n++) {
		uint32_t ds[6] = {
			n / 100000,
			(n / 10000) % 10,
			(n / 1000) % 10,
			(n / 100) % 10,
			(n / 10) % 10,
			n % 10,
		};
		if (ds[0] <= ds[1] && ds[1] <= ds[2] && ds[2] <= ds[3] && ds[3] <= ds[4] && ds[4] <= ds[5]) {
			if (ds[0] == ds[1] || ds[1] == ds[2] || ds[2] == ds[3] || ds[3] == ds[4] || ds[4] == ds[5]) {
				p1_count++;
				if ((ds[0] == ds[1] && ds[1] != ds[2])
						|| (ds[0] != ds[1] && ds[1] == ds[2] && ds[2] != ds[3])
						|| (ds[1] != ds[2] && ds[2] == ds[3] && ds[3] != ds[4])
						|| (ds[2] != ds[3] && ds[3] == ds[4] && ds[4] != ds[5])
						|| (ds[3] != ds[4] && ds[4] == ds[5])) {
					p2_count++;
				}
			}
		}
	}
    printf("Number of possible passwords: %u\n", p1_count);
    printf("Number of possible passwords (part 2): %u\n", p2_count);
	return 0;
}

File: 2019/src/bin/day4.rs
use std::env;
use std::fs::File;
use std::io;
use std::io::{BufRead, BufReader};

fn is_valid_pw(pw: &u32) -> bool {
    let ds = (
        pw / 100000,
        (pw / 10000) % 10,
        (pw / 1000) % 10,
        (pw / 100) % 10,
        (pw / 10) % 10,
        pw % 10,
    );
    // Shortcircuiting
    if !(ds.0 == ds.1 || ds.1 == ds.2 || ds.2 == ds.3 || ds.3 == ds.4 || ds.4 == ds.5) {
        return false;
    }
    if ds.0 > ds.1 || ds.1 > ds.2 || ds.2 > ds.3 || ds.3 > ds.4 || ds.4 > ds.5 {
        return false;
    }
    return true;
}

fn is_valid_pw_p2(pw: &u32) -> bool {
    let ds = (
        pw / 100000,
        (pw / 10000) % 10,
        (pw / 1000) % 10,
        (pw / 100) % 10,
        (pw / 10) % 10,
        pw % 10,
    );
    if ds.0 > ds.1 || ds.1 > ds.2 || ds.2 > ds.3 || ds.3 > ds.4 || ds.4 > ds.5 {
        return false;
    }
    return (ds.0 == ds.1 && ds.1 != ds.2)
        || (ds.0 != ds.1 && ds.1 == ds.2 && ds.2 != ds.3)
        || (ds.1 != ds.2 && ds.2 == ds.3 && ds.3 != ds.4)
        || (ds.2 != ds.3 && ds.3 == ds.4 && ds.4 != ds.5)
        || (ds.3 != ds.4 && ds.4 == ds.5);
}
fn main() -> io::Result<()> {
    let input_str: Vec<u32> = BufReader::new(
        File::open(
            &env::args()
                .nth(1)
                .expect("Incorrect number of arguments provided"),
        )
        .expect("Could not open input file"),
    )
    .lines()
    .next()
    .unwrap()
    .unwrap()
    .split('-')
    .map(|l| l.parse().unwrap())
    .collect();

    let start = input_str[0];
    let end = input_str[1];
    let count_valid = (start..=end).filter(|x| is_valid_pw(x)).count();
    println!("Number of possible passwords: {}", count_valid);
    let count_valid_p2 = (start..=end).filter(|x| is_valid_pw_p2(x)).count();
    println!("Number of possible passwords (part 2): {}", count_valid_p2);
    Ok(())
}
